### PR TITLE
docs: New tutorial on how to implement Bayesian optimization

### PR DESCRIPTION
**Pull request by mseeger**
This is the advanced developer tutorial on Bayesian optimization.

Please do give me feedback, also if something basic is missing (note there is another basic dev tutorial as well).

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/source/tutorials/dev_bayesopt/README.rst [hunk 1]
```diff
@@ -0,0 +1,22 @@
+How to Implement Bayesian Optimization
+======================================
+
+This tutorial can be seen as more advanced successor of our
+`developer tutorial <../developer/README.html>`__. It provides an overview of
+how model-based search, and in particular *Bayesian optimization*, is
+implemented in Syne, and how this code can be extended in order to fit your
```
Comment (wistuba) [inline docs/source/tutorials/dev_bayesopt/README.rst:7]: Syne Tune?

File: docs/source/tutorials/dev_bayesopt/surrogate_model.rst [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+of the complexities of Gaussian process based surrogate models:
+
+* Fantasizing is not supported
+* MCMC (or ensemble predictions) is not supported
+* Gradient-based optimization of an acquisition function is not supported, in
+  that Bayesian optimization is scoring a finite number of candidates drawn
+  at random, selecting the best
+
+The full example code is given
+`here <../../examples.html#bayesian-optimization-with-scikit-learn-based-surrogate-model>`__.
```
Comment (wistuba) [inline docs/source/tutorials/dev_bayesopt/surrogate_model.rst:231]:
```
    def predict(self, X: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
        return self.ridge.predict(X, return_std=True)
```

this does not convert `X`. Is this done automatically? Is it always done automatically?

**Review by wistuba — APPROVED** [review]
I think overall the tutorial is ok. One comment that I have is that there are a lot of references to different sources and quite a lot of reading is required before the first code is shown. An alternative way of presenting would be to first introduce the most basic requirements and a first example before diving deep into further details.

File: docs/source/tutorials/dev_bayesopt/surrogate_model.rst (hunk 2 above)
Comment (mseeger) [reply docs/source/tutorials/dev_bayesopt/surrogate_model.rst:231]: ``predict_candidates``. I insert something

**Comment by mseeger** [comment]
Thanks @wistuba . I moved the code example to the top of `surrogate_models.rst`, maybe this is better?

**Review by wistuba — APPROVED** [review]

**Pull request merged**